[quant] Add quantize_per_channel and dequantize_per_channel operators to decomposition table

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94227

Summary:
same as https://github.com/pytorch/pytorch/pull/93312, but for quantize_per_channel and dequantize_per_channel

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_q_dq_per_channel_decomposition

Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
[quant] Add quantize_per_channel and dequantize_per_channel operators to decomposition table

Summary:
same as https://github.com/pytorch/pytorch/pull/93312, but for quantize_per_channel and dequantize_per_channel

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_q_dq_per_channel_decomposition

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add quantize_per_channel and dequantize_per_channel operators to decomposition table"


Summary:
same as https://github.com/pytorch/pytorch/pull/93312, but for quantize_per_channel and dequantize_per_channel

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_q_dq_per_channel_decomposition

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vadimkantorov commented on torch/ao/quantization/_pt2e/decompositions.py:
```diff
@@ -27,3 +27,56 @@ def dequantize_per_tensor(
+    res = torch.zeros_like(input)
+
+    for i in range(input.size(0)):
+        res[i] = torch.clamp(
```
could this not be done without the loop if broadcast_right existed: https://github.com/pytorch/pytorch/issues/62148

jerryzh168 replied:
yeah probably, broadcasting is typically implicit though, how do you call broadcast_right?

vadimkantorov replied:
it doesn't exist currently in python :( created a feature request here: https://github.com/pytorch/pytorch/issues/94233

what you could do is sth like `scales = scales.unflatten(-1, scales.shape[-1:] + (1, ) * (res.dim() - scales.dim()))` (though likely not torchscriptable if it matters here)

or `torch.round(input.movedim(0, -1) / scales) + zero_points, quant_min, quant_max).movedim(-1, 0)` - although not sure what it does to preserving memory format

jerryzh168 replied:
this looks a bit complicated, I'll try transpose instead, thanks for the suggestions

vadimkantorov replied:
transpose or movedim is the cleanest, yes

vadimkantorov commented on torch/ao/quantization/_pt2e/decompositions.py:
```diff
@@ -27,3 +27,56 @@ def dequantize_per_tensor(
+        quant_max: int,
+        dtype: torch.dtype
+) -> torch.Tensor:
+    input, permute_axis_list = _permute_to_axis_zero(input, axis)
```
is it same as calling here input = input.transpose(axis, 0) and then call this again at the end of the function?

vadimkantorov replied:
So btw if this is instead dome as input.transpose(axis, -1) or even input.movedim(axis, -1) (and then the reverse at the end of the function), then regular broadcasting will work

jerryzh168 replied:
yeah feels like this is the same, I can add this as a follow up PR to improve the implementation

jerryzh168 committed:
Update on "[quant] Add quantize_per_channel and dequantize_per_channel operators to decomposition table"


Summary:
same as https://github.com/pytorch/pytorch/pull/93312, but for quantize_per_channel and dequantize_per_channel

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_q_dq_per_channel_decomposition

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add quantize_per_channel and dequantize_per_channel operators to decomposition table"


Summary:
same as https://github.com/pytorch/pytorch/pull/93312, but for quantize_per_channel and dequantize_per_channel

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_q_dq_per_channel_decomposition

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add quantize_per_channel and dequantize_per_channel operators to decomposition table"


Summary:
same as https://github.com/pytorch/pytorch/pull/93312, but for quantize_per_channel and dequantize_per_channel

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_q_dq_per_channel_decomposition

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add quantize_per_channel and dequantize_per_channel operators to decomposition table"


Summary:
same as https://github.com/pytorch/pytorch/pull/93312, but for quantize_per_channel and dequantize_per_channel

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_q_dq_per_channel_decomposition

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add quantize_per_channel and dequantize_per_channel operators to decomposition table"


Summary:
same as https://github.com/pytorch/pytorch/pull/93312, but for quantize_per_channel and dequantize_per_channel

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_q_dq_per_channel_decomposition

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vadimkantorov commented:
@jerryzh168 Sorry, can't review fully, I'm not a maintainer, just hustling around in a way

vadimkantorov commented on torch/ao/quantization/fx/_decomposed.py:
```diff
@@ -27,6 +27,40 @@ def _dequantize_per_tensor_impl(
+        input: torch.Tensor,
+        scales: torch.Tensor,
+        zero_points: torch.Tensor,
+        axis: int,
```
is `dim` here better instead of `axis` to match better pytorch conventions?

jerryzh168 replied:
yeah probably, maybe we can rename this later since this is used in existing quantize_per_channel ops right now as well

jerryzh168 committed:
Update on "[quant] Add quantize_per_channel and dequantize_per_channel operators to decomposition table"


Summary:
same as https://github.com/pytorch/pytorch/pull/93312, but for quantize_per_channel and dequantize_per_channel

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_q_dq_per_channel_decomposition

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on torch/ao/quantization/fx/_decomposed.py:
```diff
@@ -3,6 +3,40 @@
+        dtype: torch.dtype
+) -> torch.Tensor:
+    input = input.transpose(axis, -1)
+    output = torch.clamp(torch.round(input * (1.0 / scales)) + zero_points, quant_min, quant_max)
```
nit: just divide by scales?

vadimkantorov commented:
@vkuzo should this be merged or discarded?

vkuzo commented:
> @vkuzo should this be merged or discarded?

@jerryzh168  would know!

jerryzh168 commented:
haven't decided but I think it's likely going to be merged

jerryzh168 commented:
doing this in a different PR: https://github.com/pytorch/pytorch/pull/105783
